### PR TITLE
[第2組]Refactor/error handling

### PR DESCRIPTION
**Pull request by Rogue8464**
### Purpose
在專案中加入錯誤處理 (error handler) 與記錄器 (logger)，以便在錯誤處理過程中記錄相關的上下文，提升系統的可觀察性與問題排查能力。
### Details
1.新增 Logger 設定：
使用 logging 模組搭配 RotatingFileHandler，實現滾動記錄機制。
提供檔案和終端輸出的錯誤記錄功能，檔案大小與備份數量可通過專案設定調整 (LOGGER_MAX_BYTES, LOGGER_BACKUP_COUNT)。
記錄格式包括時間、名稱、日誌層級與訊息內容。
2.整合至錯誤處理：將 Logger 用於全域錯誤處理程序中，記錄錯誤詳細內容與上下文，以便日後進行問題診斷與修復。

### CODE REVIEW

**Comment by yozen0405** [comment]
這項實作完美結合了錯誤處理與記錄功能，通過設置滾動記錄機制，不僅提升了系統的可觀察性，還能有效保存錯誤的詳細上下文。整合 Logger 至全域錯誤處理程序，充分展現了系統的可維護性與專業性，為問題診斷和修復提供了強有力的支持。

**Comment by EdwardKYC** [comment]
將錯誤處理邏輯拆分為專門的例外類別，例如針對用戶操作與 API 數據驗證，這種結構清晰且便於維護。
消除了過於籠統的 try-except，使問題定位更加直觀。
新增用戶註冊表單驗證和新聞 API 架構驗證，確保數據符合預期，降低了運行時出現意外錯誤的風險。
改進新聞提取中的錯誤捕獲，避免資料處理中發生崩潰。
但我怎麼感覺你做的很少阿
日誌的層級與分類可能未按標準（INFO、WARNING、ERROR）劃分清楚。

**Comment by patrickhicetc** [comment]
在許多try except中都選擇使用`return None`而不是`raise`，建議使用後者避免其他function因為`return None`而產生問題。
`logger`似乎清一色全都是`logger.error`，建議有些較輕微的可以用`logger.warning`

**Comment by Steven0811** [comment]
這份PR有許多問題，首先是logger的部分，建議可以加入使用者操作與系統內部函式調用等日誌紀錄，另外在logging的等級劃分上也有問題，error等級通常用於無法讓系統繼續執行的錯誤，使用者操作失誤和其他較輕微的失誤可以使用warning或info就好
另外針對各個函式中error handling的部分，在出現問題時你用了很多```return None```，但那很有可能在造成其他函式調用的問題，建議可使用```raise```，還有我看不到你有使用Sentry來追蹤錯誤，雖然你有該Sentry的DSN，但我沒看到你在錯誤發生時有使用```capture_exception```
不過還是有一些做得好的地方，你在使用try except的時候包含的範圍很小，像是
```python
try:
    title = soup.find("h1", class_="article-content__title").text
except Exception as e:
    title = "無法取得標題"
    logger.error(f"Unable to fetch news title:{e}",exc_info=True,)
try:
    time = soup.find("time", class_="article-content__time").text
except Exception as e:
    time = "無法取得時間"
    logger.error(f"Unable to fetch news time:{e}",exc_info=True,)
try:
    content_section = soup.find("section", class_="article-content__editor")
except Exception as e:
    content_section = "無法取得內文"
    logger.error(f"Unable to fetch news content:{e}",exc_info=True,)
```
儘管如此，全部使用```except Exception as e```並不好，因為Exception有可能同時包含客戶端錯誤和系統錯誤，這會讓logger的階級變得模糊

**Comment by ptchen314** [comment]
PR沒有正常遵循MD格式，但這問題不大。
try except的部分有部分內容需要修正，若非必要情況請不要用as e，這樣到最後會很容易找不到錯誤的部分。
return None這部分有概率會抽風，建議使用raise XXX來保證fastAPI不抽風

**Comment by mlcp29th-hi** [comment]
I appreciate the work done in this PR, particularly the increased focus on error handling throughout the codebase. Here are a few suggestions and considerations for improving error handling:

1. **Specific Exception Types**: 
   - In many places, you are catching a generic `Exception`. It would be more beneficial to catch specific exceptions where possible (e.g., `requests.exceptions.RequestException`, `ValueError`, etc.). This will allow you to have more fine-grained control over error handling and make debugging easier.

2. **Consistent Response Handling**: 
   - When an error occurs, returning `None` (or `None` with a structured error message) can lead to inconsistencies in handling responses throughout your application. Consider standardizing error responses through an error handling middleware or returned objects that encapsulate both the data and error state.

3. **Logging Best Practices**:
   - While using the logger to capture errors is great, ensure that sensitive information (like user passwords or tokens) is never logged. If you need to include method inputs in log messages, sanitize them first.

4. **Before Returning Assuming Failure**: 
   - In the `_perform_request` function, if a request fails, returning `None` might not inform the caller sufficiently. Consider throwing a custom exception or returning an error object that includes both error type and message, which would provide more context.

5. **Handling Unforeseen Outputs**:
   - In functions like `sum_up_news`, when parsing JSON results, add a check to see if `result` exists before trying to access `result["影響"]` or `result["原因"]`. If `result` is `None` or does not have those keys, it could raise a `KeyError`.

6. **Error Handling on Database Operations**:
   - Consider introducing a dedicated error handler for your database operations. Logging errors while trying to execute statements is acceptable, but it’s also critical to validate the state of your database connection before making these calls and handle potential connection issues.

7. **Using Try-Except Sparingly**:
   - You can minimize the footprint of try-except blocks by structuring code in a way that the highest level of abstraction handles the errors. For example, if multiple operations can fail, a single try-except around the method call may suffice rather than each individual operation.

8. **Use of `exc_info=True`**:
   - You’ve done a good job adding `exc_info=True` in your logger statements, which helps in stack tracing. Just ensure that this is consistent across all log messages that capture exceptions.

By addressing the concerns above, you can improve the maintainability and robustness of the code you are developing. Overall, great job on enhancing the error handling mechanisms in your codebase!

**Comment by Yi-Cheng-Wang** [comment]
## Advantages

1. **Clear Error Feedback**:  
   Using `catch` blocks to capture exceptions and provide specific error messages helps users understand what went wrong.  

2. **State Management**:  
   Error handling can be integrated with the application's state management system (e.g., Pinia), allowing the UI to reflect the current loading and error states, thereby improving the user experience.  

3. **Prevention of Application Crashes**:  
   Capturing exceptions prevents the application from crashing due to unhandled errors, increasing overall stability.  

4. **Extensibility**:  
   Error handling logic can be extended as needed, for example, by adding logging or error reporting features for deeper issue analysis.  

## Disadvantages

1. **Redundant Error Handling**:  
   Writing `try...catch` blocks for every asynchronous operation can result in verbose and difficult-to-maintain code.  

2. **Vague Error Messages**:  
   If error messages fail to clearly describe the issue, users might become confused, especially when the errors originate from the backend.  

3. **Performance Impact**:  
   Excessive error handling, particularly in high-frequency operations, might impact performance.  

4. **Lack of Unified Error Handling Strategy**:  
   If different components or modules employ inconsistent error-handling approaches, it may lead to an overall lack of cohesion, complicating maintenance.  

## Summary

Error handling is an essential part of application development. While it has its advantages and drawbacks, a well-designed and implemented strategy can maximize user experience and application stability. It's recommended to adopt a unified error-handling approach and abstract common logic to minimize code duplication and improve maintainability.

**Comment by percykugood** [comment]
大致上寫的很好，但是建議根據不同的運行環境設置不同的日誌級別和輸出方式，以避免在環境中產生過多的日誌，影響性能。含有建議減少過於寬泛的異常處理（如直接使用 Exception），針對特定的錯誤類型（如 KeyError, ValueError, HTTPError）來寫專門的處理邏輯，避免掩蓋掉潛在的問題。

**Comment by Diego0923** [comment]
成功完成了要求good，但建議可以做些改動
日誌分級需依INFO/WARNING/ERROR標準調整，提升可讀性；
避免return None，改用raise以增強錯誤處理一致性；
細化Exception類型，優化除錯效率。

File: backend/src/user/service.py [hunk 1]
```diff
@@ -6,30 +6,45 @@
...
     return password_context.verify(p1, p2)
 
 def check_user_password_is_correct(db, user_input, password):
-    user = db.query(User).filter(User.username == user_input).first()
-    if not verify_hashed_password(password, user.hashed_password):
+    try:
+        user = db.query(User).filter(User.username == user_input).first()
+        user_hashed_password = user.hashed_password
+    except Exception as e:
+
```
Comment (dee9c-git) [inline backend/src/user/service.py:19]: Nitpicking: odd \n

**Review by New9c — COMMENTED** [review]
I think the code isn't too bad, so I'll tell you about MD:
```## Purpose``` 
will result in:
## Purpose
```##Purpose``` 
will result in:
##Purpose
Add a space.

You can use **Preview** to see how it will look
```You can use **Preview** to see how it will look```

**Pull request merged**